vllm/v1/executor/ray_executor.py

```diff
@@ -0,0 +1,356 @@
+import os
+from collections import defaultdict
+from itertools import islice, repeat
+from typing import (TYPE_CHECKING, Any, Callable, Dict, List, Optional, Tuple,
+                    Type)
+
+import vllm.envs as envs
+from vllm.config import VllmConfig
+from vllm.logger import init_logger
+from vllm.utils import get_distributed_init_method, get_ip, get_open_port
+from vllm.v1.executor.ray_utils import RayWorkerWrapper, ray
+from vllm.v1.outputs import ModelRunnerOutput
+from vllm.worker.worker_base import WorkerBase
+
+if ray is not None:
+    from ray.util.scheduling_strategies import PlacementGroupSchedulingStrategy
+
+if TYPE_CHECKING:
+    from ray.util.placement_group import PlacementGroup
+
+logger = init_logger(__name__)
+
+
+class RayExecutor:
+
+    def __init__(self, vllm_config: VllmConfig) -> None:
+        self.vllm_config = vllm_config
+        self.model_config = vllm_config.model_config
+        self.cache_config = vllm_config.cache_config
+        self.lora_config = vllm_config.lora_config
+        self.load_config = vllm_config.load_config
+        self.parallel_config = vllm_config.parallel_config
+        self.scheduler_config = vllm_config.scheduler_config
+        self.device_config = vllm_config.device_config
+        self.speculative_config = vllm_config.speculative_config
+        self.prompt_adapter_config = vllm_config.prompt_adapter_config
+        self.observability_config = vllm_config.observability_config
+        self._init_executor()
+
+    def _init_executor(self) -> None:
+        self.forward_dag: Optional[ray.dag.CompiledDAG] = None
+        placement_group = self.parallel_config.placement_group
+
+        # Disable Ray usage stats collection.
+        ray_usage = os.environ.get("RAY_USAGE_STATS_ENABLED", "0")
+        if ray_usage != "1":
+            os.environ["RAY_USAGE_STATS_ENABLED"] = "0"
+
+        # Create the parallel GPU workers.
+        self._init_workers_ray(placement_group)
+
+    def _init_workers_ray(self, placement_group: "PlacementGroup",
+                          **ray_remote_kwargs):
+        if (self.parallel_config.tensor_parallel_size == 1
+                and self.parallel_config.pipeline_parallel_size == 1):
+            # For single GPU case, we use a ray worker with constrained memory.
+            num_gpus = self.cache_config.gpu_memory_utilization
+        else:
+            # Otherwise, the ray workers are allocated with a full GPU.
+            num_gpus = 1
```

Comment:
In the single GPU case with num_gpus=0.9 for example, will `torch.cuda.mem_get_info()` only see 0.9 of the total memory? The reason for this question is in GPU worker we use the following logic to determine the KV blocks:

```python
_, total_gpu_memory = torch.cuda.mem_get_info()
...
available_kv_cache_memory = (
    total_gpu_memory * self.cache_config.gpu_memory_utilization -
    peak_memory)
```

So if `total_gpu_memory` is already 0.9, then we will only use 0.81 of the total memory.

Reply:
`num_gpus` should not affect the memory obtained from `torch.cuda.mem_get_info()`. For example, when running this script:

```
import ray

ray.init(num_gpus=1)

@ray.remote
class FractionalGPUActor:
    def ping(self):
        import torch
        gpu_id = ray.get_runtime_context().get_accelerator_ids()["GPU"]
        mem_info = torch.cuda.mem_get_info()
        print(f"gpu_id: {gpu_id}, mem_info: {mem_info}")

fractional_gpu_actors = [FractionalGPUActor.options(num_gpus=0.1*(i+1)).remote() for i in range(3)]
# Ray tries to pack GPUs if possible.
[ray.get(fractional_gpu_actors[i].ping.remote()) for i in range(3)]
```

We got:
```
(FractionalGPUActor pid=3762944) gpu_id: ['0'], mem_info: (23382261760, 23583784960)
(FractionalGPUActor pid=3762941) gpu_id: ['0'], mem_info: (23184015360, 23583784960)
(FractionalGPUActor pid=3762946) gpu_id: ['0'], mem_info: (22985703424, 23583784960)
```

The total memory is the same.